File: C++/Arrays/TwoSum.cpp
//Given an array of integers nums and an integer target, return the indices i and j such that nums[i] + nums[j] == target and i != j.
//You may assume that every input has exactly one pair of indices i and j that satisfy the condition.
//Return the answer with the smaller index first.
//Time complexity O(n) || Space complexity O(n)

#include <iostream>
#include <unordered_map>
#include <vector>
using namespace std;

int main()
{
    int size;
    cout << "Enter the size of array: ";
    cin >> size;

    vector<int> nums(size);
    for (int i = 0; i < size; i++)
    {
        cout << "Enter " << i + 1 << "th element of array: ";
        cin >> nums[i];
    }

    int target;
    cout << "Enter target: ";
    cin >> target;
    
    unordered_map<int, int> mp;
    for (int i = 0; i < size; i++)
    {
        int complement = target - nums[i];
        if (mp.find(complement) != mp.end()) //checks whether the complement exists in the mp
        {
            int index1 = mp[complement];
            int index2 = i;
            if (index1 < index2) // print the smaller index first
            {
                cout << index1 << " " << index2 << endl;
            }
            else
            {
                cout << index2 << " " << index1 << endl;
            }
            break;
        }
        mp[arr[i]] = i; // store the index of the current element
    }

    return 0;
}

File: Python/Arrays/Two-Sum.py
#Given an array of integers nums and an integer target, return the indices i and j such that nums[i] + nums[j] == target and i != j.

class Solutions_twoSum:
    def twoSum(self, nums, target):
        indices = {}

        for index, number in enumerate(nums):
            indices[number] = index
        
        for index, number in enumerate(nums):
            diff = target - number
            if diff in indices and indices[diff] != index:
                return [index, indices[diff]]

test_case = Solutions_twoSum()
print(test_case.twoSum(nums = [1, 3, 5, 7, 9, 4], target = 7))